asmcomp/linscan.ml

```diff
@@ -41,6 +47,26 @@ let rec insert_interval_sorted i = function
   | j :: _ as il when j.iend <= i.iend -> i :: il
   | j :: il -> j :: insert_interval_sorted i il
 
+(* Note that we do not call [Interval.remove_expired_ranges] in
+   [release_expired_spilled], unlike in the rest of the [remove_expired_*]
+   functions.
+
+   [remove_expired_ranges] discards already handled live ranges in each interval
+   to make [Interval.is_live] and [Interval.overlap] more efficient.
+
+   As we do not call either of these functions in [release_expired_spilled] we
+   do not need to call [remove_expired_ranges]. *)
+
+let rec release_expired_spilled ci pos = function
+    i :: il when i.iend >= pos ->
+      i :: release_expired_spilled ci pos il
```

Comment:
Wouldn't it be better to keep ci_spilled sorted by increasing i.iend? That's more work when inserting, but we avoid scanning it entirely for every interval.

Reply:
Right, this is what I had done in a previous version. But since we need to iterate through the rest of the list anyway to collect the stack slots to be reused, it doesn't look like we would gain a lot here (concretely, we would avoid one branch for each of the expired intervals) but we would still pay the cost at insertion time.

Do you agree?

Reply:
I'm not sure to agree :-)

If ci_spilled is kept sorted by increasing i.iend, then for every new interval, we only need to scan a **prefix** of that list in order to release the corresponding slots (the rest of the list remains untouched).

Reply:
Actually, ci_spilled could be represented by a set of integer pairs (sorted by lexicographic ordering)  `(iend, ss)` (no need to keep the intervals themselves).  For each interval, one would iterate as long as the smallest item can be released (and thus removed from this set).  (A better data structure would be a heap, actually.)

Reply:
Right, good idea! (I had only thought keeping the intervals sorted by _decreasing_ `i.iend` which doesn't quite work).

I pushed a commit keeping the intervals sorted as suggested.

On the other hand, I don't quite understand the comment about keeping the pairs `(iend, ss)` lexicographically ordered. Is it about getting rid of the `release_expired_spilled` operation altogether and instead extract/update the stack slot to be reused directly when spilling a new interval in `allocate_stack_slot`?

Reply:
> On the other hand, I don't quite understand the comment about keeping the pairs (iend, ss) lexicographically ordered. Is it about getting rid of the release_expired_spilled operation altogether and instead extract/update the stack slot to be reused directly when spilling a new interval in allocate_stack_slot?

My proposal is just to use a set of `int*int` instead of a sorted list of `int*int`.  We can both insert an element and remove the smallest one in (amortized) logarithmic time; while the sorted list representation requires a linear pass (but removal would be fastest).  But, agreed, it's really not clear that it would improve performance in practice.

Reply:
Thanks for the explanation. I tried it and indeed it gives a nice speedup: with this change, the benchmark timings for this PR are even with those of `trunk` (which only does the more naïve spilling method).